src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.EventDispatch.cs

```diff
@@ -79,37 +79,38 @@ private Task ReturnErrorResponse(string detailedMessage)
             : Task.CompletedTask;
     }
 
-    internal async Task SetNotFoundResponseAsync(string baseUri, NotFoundEventArgs args)
+    internal void SetNotFoundWhenResponseNotStarted()
     {
-        if (_httpContext.Response.HasStarted ||
-            // POST waits for quiescence -> rendering the NotFoundPage would be queued for the next batch
-            // but we want to send the signal to the renderer to stop rendering future batches -> use client rendering
-            HttpMethods.IsPost(_httpContext.Request.Method))
-        {
-            if (string.IsNullOrEmpty(_notFoundUrl))
-            {
-                _notFoundUrl = GetNotFoundUrl(baseUri, args);
-            }
-            var defaultBufferSize = 16 * 1024;
-            await using var writer = new HttpResponseStreamWriter(_httpContext.Response.Body, Encoding.UTF8, defaultBufferSize, ArrayPool<byte>.Shared, ArrayPool<char>.Shared);
-            using var bufferWriter = new BufferedTextWriter(writer);
-            HandleNotFoundAfterResponseStarted(bufferWriter, _httpContext, _notFoundUrl);
-            await bufferWriter.FlushAsync();
-        }
-        else
+        _httpContext.Response.StatusCode = StatusCodes.Status404NotFound;
+
+        // When the application triggers a NotFound event, we continue rendering the current batch.
+        // However, after completing this batch, we do not want to process any further UI updates,
+        // as we are going to return a 404 status and discard the UI updates generated so far.
+        SignalRendererToFinishRendering();
+    }
+
+    internal async Task SetNotFoundWhenResponseHasStarted()
+    {
+        if (string.IsNullOrEmpty(_notFoundUrl))
         {
-            _httpContext.Response.StatusCode = StatusCodes.Status404NotFound;
+            var baseUri = $"{_httpContext.Request.Scheme}://{_httpContext.Request.Host}{_httpContext.Request.PathBase}/";
+            _notFoundUrl = GetNotFoundUrl(baseUri, NotFoundEventArgs);
         }
+        var defaultBufferSize = 16 * 1024;
+        await using var writer = new HttpResponseStreamWriter(_httpContext.Response.Body, Encoding.UTF8, defaultBufferSize, ArrayPool<byte>.Shared, ArrayPool<char>.Shared);
+        using var bufferWriter = new BufferedTextWriter(writer);
+        HandleNotFoundAfterResponseStarted(bufferWriter, _httpContext, _notFoundUrl);
+        await bufferWriter.FlushAsync();
 
         // When the application triggers a NotFound event, we continue rendering the current batch.
         // However, after completing this batch, we do not want to process any further UI updates,
         // as we are going to return a 404 status and discard the UI updates generated so far.
         SignalRendererToFinishRendering();
     }
 
-    private string GetNotFoundUrl(string baseUri, NotFoundEventArgs args)
+    private string GetNotFoundUrl(string baseUri, NotFoundEventArgs? args)
     {
-        string path = args.Path;
+        string? path = args?.Path;
```

Comment:
The error `"The NotFoundPage route must be specified or re-execution middleware has to be set to render NotFoundPage when the response has started."` could be simply "The NotFoundPage route must be specified or re-execution middleware has to be set to render not found content." now, when we don't support not found fragment. Also, there should be `nameof`.